### PR TITLE
fix: Run PreWork once, even for parallel analyzers

### PR DESCRIPTION
**Pull request by ptrus**
Fixes: https://github.com/oasisprotocol/nexus/issues/803

### CODE REVIEW

File: analyzer/block/block.go [hunk 1]
```diff
@@ -39,7 +39,9 @@ const (
 // block based analyzer.
 type BlockProcessor interface {
 	// PreWork performs tasks that need to be done before the main processing loop starts.
+	// When running in parallel mode, this method will only be called once.
```
Comment (pro-wh) [inline analyzer/block/block.go:42]: more info: when running in parallel mode with multiple analyzer instances, only one of those instances will have PreWork called. so don't use it to initialize any fields of the processor. only use it for shared state, such as fixing up the database

**Review by pro-wh — APPROVED** [review]
good, I've wanted this

**Comment by pro-wh** [comment]
o sorry I gotta ask a question first

File: analyzer/block/block.go [hunk 2]
```diff
@@ -316,16 +333,6 @@ func (b *blockBasedAnalyzer) Start(ctx context.Context) {
 		to = b.blockRange.To
```
Comment (pro-wh) [inline analyzer/block/block.go:333]: does any prework depend on this

**Review by pro-wh — APPROVED** [review]
ok nvm it calls PreWork on both one fast sync analyzer and the slow sync analyzer. seems fine as the PreWork internally branches on b.slowSync. everybody just keep that in mind

File: analyzer/block/block.go (hunk 2 above)
Comment (ptrus) [reply analyzer/block/block.go:333]: Nope

File: cmd/analyzer/analyzer.go [hunk 3]
```diff
@@ -694,16 +694,31 @@ func (a *Service) Start() {
...
 			fastSyncWg[an.SyncTag] = wg
+
+			// This is the first fast-sync analyzer with this SyncTag, so PreWork should be run.
+			// This relies on the assumption that each distinct 'type' of fast-sync analyzer uses a unique `SyncTag`.
+			// This assumption holds true at the moment.
+			runPreWork = true
 		}
 		wg.Add(1)
-		go func(an SyncedAnalyzer) {
+		go func(an SyncedAnalyzer, runPreWork bool) {
```
[the remark this answers is not in the document]
Comment (ptrus) [reply cmd/analyzer/analyzer.go]: Thanks, updated for consistency with `wg`. Although, for me the explicit goroutine param seems a bit cleaner to read.

File: analyzer/block/block.go (hunk 2 above)
Comment (pro-wh) [reply analyzer/block/block.go:333]: kool

File: cmd/analyzer/analyzer.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pro-wh) [reply cmd/analyzer/analyzer.go]: oh ok, thanks

**Pull request merged**